packages/exo/src/exo-tools.js

```diff
@@ -6,59 +6,156 @@ import {
  * A method guard, for inclusion in an interface guard, that enforces only that
  * all arguments are passable and that the result is passable. (In far classes,
- * "any" means any *passable*.) This is the least possible enforcement for a
- * method guard, and is implied by all other method guards.
+ * "any" means any *passable*.) This is the least possible non-raw
+ * enforcement for a method guard, and is implied by all other
+ * non-raw method guards.
+ */
+const PassableMethodGuard = M.call().rest(M.any()).returns(M.any());
+
+/**
+ * @typedef {object} RedactConfig
+ * @property {number} declaredLen
+ * @property {boolean} hasRestArgGuard
+ * @property {boolean} needRedaction
+ * @property {Pattern} paramsPattern
+ * @property {boolean} rawRestArgs
+ * @property {number[]} redactedIndices
  */
-const MinMethodGuard = M.call().rest(M.any()).returns(M.any());
 
 /**
  * @param {Passable[]} syncArgs
- * @param {MethodGuardPayload} methodGuardPayload
+ * @param {RedactConfig} redactConfig
  * @param {string} [label]
  * @returns {Passable[]} Returns the args that should be passed to the
  * raw method
  */
-const defendSyncArgs = (syncArgs, methodGuardPayload, label = undefined) => {
-  const { argGuards, optionalArgGuards, restArgGuard } = methodGuardPayload;
-  const paramsPattern = M.splitArray(
-    argGuards,
-    optionalArgGuards,
-    restArgGuard,
-  );
-  mustMatch(harden(syncArgs), paramsPattern, label);
-  if (restArgGuard !== undefined) {
+const defendSyncArgs = (syncArgs, redactConfig, label = undefined) => {
+  const {
+    declaredLen,
+    hasRestArgGuard,
+    needRedaction,
+    paramsPattern,
+    rawRestArgs,
+    redactedIndices,
+  } = redactConfig;
+
+  // Only manipulate the original args if our params pattern didn't contain any
+  // raw values.
+  let redactedArgs = syncArgs;
+  if (rawRestArgs) {
+    // Don't harden the rest args.
+    redactedArgs = syncArgs.slice(0, Math.min(syncArgs.length, declaredLen));
+  } else if (needRedaction) {
+    // Copy the arguments array, avoiding hardening the redacted ones (which are
+    // trivially matched).
+    redactedArgs = [...syncArgs];
+  }
+
+  for (const i of redactedIndices) {
+    if (i >= redactedArgs.length) {
+      break;
+    }
+    redactedArgs[i] = REDACTED_RAW_ARG;
+  }
+
+  mustMatch(harden(redactedArgs), paramsPattern, label);
+
+  if (hasRestArgGuard) {
     return syncArgs;
   }
-  const declaredLen =
-    argGuards.length + (optionalArgGuards ? optionalArgGuards.length : 0);
   if (syncArgs.length <= declaredLen) {
     return syncArgs;
   }
   // Ignore extraneous arguments, as a JS function call would do.
   return syncArgs.slice(0, declaredLen);
 };
 
+/**
+ * @param {MethodGuardPayload} methodGuardPayload
+ */
+const redactRawArgs = methodGuardPayload => {
+  const {
+    argGuards,
+    optionalArgGuards = [],
+    restArgGuard,
+  } = methodGuardPayload;
+
+  const rawArgGuards = [...argGuards, ...optionalArgGuards];
+
+  let needRedaction = false;
```

Comment:
this state is implied by `redactedIndices`. consider omitting this and down below returning,
```js
    needsRedaction: redactedIndicies.length > 0,
```
Though since it's implied by the length of that array, why pass it around at all?

Reply:
Omitted and using `redactedIndices` in its place.